### ISSUE TITLE
[question] getting started/basics

### ISSUE THREAD

**Issue by geoHeil** [benedekrozemberczki/pytorch_geometric_temporal#59]
Many thankgs for the great library!

However I am just getting started and have several questions:

- how can I ensure that the GPU/device is used and not the CPU? By default the examples seem to run on the CPU. I figured out that I need to write: 
```python
%%time

from tqdm import tqdm
model = RecurrentGCN(node_features = 4)
model.cuda()
# How to use CUDA? How to move data to GPU?
#train_dataset.cuda()
optimizer = torch.optim.Adam(model.parameters(), lr=0.01)
model.train()
for epoch in tqdm(range(200)):
    cost = 0
    for time, snapshot in enumerate(train_dataset):
        y_hat = model(snapshot.x.cuda(), snapshot.edge_index.cuda(), snapshot.edge_attr.cuda())
        cost = cost + torch.mean((y_hat-snapshot.y.cuda())**2)
    cost = cost / (time+1)
    cost.backward()
    optimizer.step()
    optimizer.zero_grad() 
```
But always calling `x.cuda()` seems to be really tedious. Is it possible to manage this in a more centralized manner?

- how do I control the batch size? I have the feeling you call this snapshots? The `_get_snapshot` seems to iterate these (at least for the 2 examples from the documentation) however, I do not understand how multiple snaphshots / batches would be created or registered to iterate on? Or is batch_size and snapshot not the same concept? I observe that: `self.t = self.t + 1` you are simply counting up the (discrete) timesteps - but this assumes discrete timesteps - even for `DynamicGraphTemporalSignal`.  How would contionous timestamps need to be handled? Others like Stellar - https://stellargraph.readthedocs.io/en/stable/demos/link-prediction/ctdne-link-prediction.html - would use  a Continuous-Time Dynamic Network Embeddings (CTDNE) fur such purposes.

**Comment by benedekrozemberczki** [comment]
Dear @geoHeil,,

It is nice to hear that you like the library! 

A, You can transfer the snapshots because those are PyTorch Geometric Dataset Objects. Given that temporal snapshots can be large it is not evident that transfering all the snapshots in advance would be the way to go.

B, Batching is more complicated than snapshots. Consider this:

1, You could batch though time (accumulate loss in windows).
2, Batch on receptive fields that you sample (diagonal batching trick). This requires taking the same receptive field across time steps - for this you have to create a sampling logic on your own which is non trivial.

C, Representing continuous time is also a non trivial problem:
   - Features can change.
   - The node set and edge set can change.
 We are actively working on a way to handle time difference which are non constant. If you have a specific design idea feel free to get involved.

Bests,

Benedek

**Comment by geoHeil** [comment]
Many thanks for the quick reply.

I think A is clear now. I would hope that when using https://www.pytorchlightning.ai/ in the future this gets managed more automatically.

B: agreed. And I would opt for 1  i.e. have sliding window batches. Do you have an example implementation for this? So far it looked like what I have found on your web page was always using a single batch with multiple snapshots.

C: let me experiment with discrete time a bit more - and then I will try to get involved.

Basically my dataset looks like this:

```
import pandas as pd

df = pd.DataFrame({'person_1':[], 'person_2':[], 'time':[], 'lat':[], 'long':[], 'hobby':[]})
display(df)
```

and I want to perform link prediction: I ant to suggest new links/friendships for similar hobbies - ideally performing these not too far away from ones own locations.

May I add one more question: regarding handling of space are there some methods you suggest? Or can these coordinates simply be kept as regular features on the edge?

**Comment by benedekrozemberczki** [comment]
Normalizing the distances to a certain reference point might help. Just an idea.
